src/analysis/calculus/deriv.lean

```diff
@@ -0,0 +1,610 @@
+  tendsto (λ x', ∥x' - x∥⁻¹ * ∥f x' - f x - (x' - x) • f'∥) (nhds_within x s) (nhds 0) :=
+has_fderiv_at_filter_iff_tendsto
+
+theorem has_deriv_at_iff_tendsto : has_deriv_at f f' x ↔
+  tendsto (λ x', ∥x' - x∥⁻¹ * ∥f x' - f x - (x' - x) • f'∥) (nhds x) (nhds 0) :=
+has_fderiv_at_filter_iff_tendsto
+
+theorem has_deriv_at_filter.mono (h : has_deriv_at_filter f f' x L₂) (hst : L₁ ≤ L₂) :
+  has_deriv_at_filter f f' x L₁ :=
+has_fderiv_at_filter.mono h hst
+
+theorem has_deriv_within_at.mono (h : has_deriv_within_at f f' t x) (hst : s ⊆ t) :
+  has_deriv_within_at f f' s x :=
+has_fderiv_within_at.mono h hst
+
+theorem has_deriv_at.has_deriv_at_filter (h : has_deriv_at f f' x) (hL : L ≤ nhds x) :
+  has_deriv_at_filter f f' x L :=
+has_fderiv_at.has_fderiv_at_filter h hL
+
+theorem has_deriv_at.has_deriv_within_at
+  (h : has_deriv_at f f' x) : has_deriv_within_at f f' s x :=
+has_fderiv_at.has_fderiv_within_at h
+
+lemma has_deriv_within_at.differentiable_within_at (h : has_deriv_within_at f f' s x) :
+  differentiable_within_at 𝕜 f s x :=
+has_fderiv_within_at.differentiable_within_at h
+
+lemma has_deriv_at.differentiable_at (h : has_deriv_at f f' x) : differentiable_at 𝕜 f x :=
+has_fderiv_at.differentiable_at h
+
+@[simp] lemma has_deriv_within_at_univ : has_deriv_within_at f f' univ x ↔ has_deriv_at f f' x :=
+has_fderiv_within_at_univ
+
+theorem has_deriv_at_unique
+  (h₀ : has_deriv_at f f₀' x) (h₁ : has_deriv_at f f₁' x) : f₀' = f₁' :=
+smul_right_one_eq_iff.mp $ has_fderiv_at_unique h₀ h₁
+
+lemma has_deriv_within_at_inter' (h : t ∈ nhds_within x s) :
+  has_deriv_within_at f f' (s ∩ t) x ↔ has_deriv_within_at f f' s x :=
+has_fderiv_within_at_inter' h
+
+lemma has_deriv_within_at_inter (h : t ∈ nhds x) :
+  has_deriv_within_at f f' (s ∩ t) x ↔ has_deriv_within_at f f' s x :=
+has_fderiv_within_at_inter h
+
+lemma differentiable_within_at.has_deriv_within_at (h : differentiable_within_at 𝕜 f s x) :
+  has_deriv_within_at f (deriv_within f s x) s x :=
+have h' : ∃ f', has_deriv_within_at f f' s x,
+  from let ⟨f', h⟩ := h in ⟨_, has_fderiv_within_at_iff_has_deriv_within_at.mp h⟩,
+by dunfold deriv_within; rw dif_pos h'; exact classical.some_spec h'
+
+lemma differentiable_at.has_deriv_at (h : differentiable_at 𝕜 f x) :
+  has_deriv_at f (deriv f x) x :=
+have h' : ∃ f', has_deriv_at f f' x,
+  from let ⟨f', h⟩ := h in ⟨_, has_fderiv_at_iff_has_deriv_at.mp h⟩,
+by dunfold deriv; rw dif_pos h'; exact classical.some_spec h'
+
+lemma has_deriv_at.deriv (h : has_deriv_at f f' x) : deriv f x = f' :=
+has_deriv_at_unique h.differentiable_at.has_deriv_at h
+
+lemma has_deriv_within_at.deriv_within
+  (h : has_deriv_within_at f f' s x) (hxs : unique_diff_within_at 𝕜 s x) :
+  deriv_within f s x = f' :=
+hxs.eq_deriv _ h.differentiable_within_at.has_deriv_within_at h
+
+lemma fderiv_within_deriv_within (hxs : unique_diff_within_at 𝕜 s x) :
+  (fderiv_within 𝕜 f s x : 𝕜 → F) 1 = deriv_within f s x :=
```

Comment:
Wouldn't it make sense to /define/ `deriv_within f s x` as `(fderiv_within 𝕜 f s x : 𝕜 → F) 1`, to avoid the unique differential assumption here? Currently, there is a choice to define `deriv_within` and a choice to define `fderiv_within`, and there is not guarantee that these choices coincide...

Reply:
Ok, I also defined `deriv f x` as `fderiv 𝕜 f x 1`.